### ISSUE TITLE
Use type context to infer type of empty collections in and/or expressions

### ISSUE THREAD

**Issue by zunger-humu** [python/mypy#6572]
Underlying issue: for any generic type, mypy considers GenType[<nothing>] and GenType[X] to be incompatible classes. These should instead be considered to be of type GenType[X]. (Observed with mypy 0.670 / Python 3.6, no flags passed to mypy)

To illustrate why, consider two common Python idioms where this happens. One is in a function declaration, where None is passed as a default value rather than a class-scoped mutable variable:
```
def fn(arg: Set[str] = None) -> None:
    arg = arg or set()
    ...
```
This is reported as an error, "Incompatible types in assignment (expression has type "Union[Set[str], Set[<nothing>]]", variable has type "Optional[Set[str]]")"

Likewise, if an abstract superclass declares a method which yields an iterator,
```
def abstractMethod(self) -> Iterator[...]:
    pass
```

Because the abstract declaration passes, it is actually returning an optional iterator (which mypy correctly understands). Callers of the abstract method then cope with this easily:
```
for value in (thing.abstractMethod() or []):
   ...
```
which yields the same error as above, or
```
yield from (thing.abstractMethod() or [])
```
which fails with "error: "yield from" can't be applied to "Union[Iterator[type], List[<nothing>]]".

The expected behavior in all cases is no error: empty iterables should be considered to be valid instances of iterables of any type, without further annotation.

**Comment by ilevkivskyi** [comment]
`List[<nothing>]` is not a subtype of `List[X]` because it is invariant, for example
```python
x = []  # mypy would complain here, but assume we ignore it, so 'x' is 'List[<nothing>]'

y: List[str] = x  # this is dangerous
y.append('boom')

z: List[int] = x
z[0] + 42  # TypeError here
```
Instead mypy should use the outer type context to infer the correct type of an empty collection. I will edit the title accordingly.

**Comment by ilevkivskyi** [comment]
Also this is low priority since it may be tricky to implement one can instead simply write:
```python
def func(lst: Optional[List[int]] = None) -> None:
    if lst is None:
        lst = []
```
(which is also more readable IMO)

**Comment by JukkaL** [comment]
It might be reasonable to make `List[<nothing>]` a subtype of `List[X]` though (#6613).

**Comment by AlexWaygood** [comment]
[Cannot reproduce](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=d9caf0c87e0170ba40c57658ca3cbe46) any of these false-positive errors on mypy 0.941; mypy seems to cope with these fine now.
